swarm/storage/filestore.go

```diff
@@ -96,3 +96,41 @@ func (f *FileStore) Store(ctx context.Context, data io.Reader, size int64, toEnc
 func (f *FileStore) HashSize() int {
 	return f.hashFunc().Size()
 }
+
+// Public API. This endpoint returns all chunk hashes (only) for a given file
+func (f *FileStore) GetAllReferences(ctx context.Context, data io.Reader, toEncrypt bool) (addrs []Address, err error) {
+	addrs = make([]Address, 0)
+	// create a special kind of putter, which only will store the references
+	putter := &HashExplorer{
+		hasherStore: NewHasherStore(f.ChunkStore, f.hashFunc, toEncrypt),
+		References:  make([]Reference, 0),
+	}
+	// do the actual splitting anyway, no way around it
+	_, _, err = PyramidSplit(ctx, data, putter, putter)
+	if err != nil {
+		return nil, err
+	}
+	// collect all references
+	for _, ref := range putter.References {
+		addrs = append(addrs, Address(ref))
+	}
+	return addrs, nil
+}
+
+// HashExplorer is a special kind of putter which will only store chunk references
+type HashExplorer struct {
+	*hasherStore
+	References []Reference
+}
+
+// HashExplorer's Put will add just the chunk hashes to its `References`
+func (he *HashExplorer) Put(ctx context.Context, chunkData ChunkData) (Reference, error) {
+	// Need to do the actual Put, which returns the references
+	ref, err := he.hasherStore.Put(ctx, chunkData)
```

Comment:
Oh my, it's pretty clumsy having to hash everything twice... I wonder why the pyramidsplitter only returns the data, not the reference.

Reply:
@nolash not *everything* is hashed twice - and the points in time for the requests are separated.

We use `Store` when we actually really store a data structure on swarm. Then we use the "conventional" `PyramidSplit`, and having all references for the data structure is (currently) not needed and useless.

We use `GetAllReferences` for debugging, when we ask a node "do you actually (still?) have a chunk with hash `abc123` in your store? *Then* It will be hashed again (but only by the *checking* node) to get all references for the given data structure, so it is at another point in time, and optionally for user-selected data structures.

Is the use case clear now?